core/sql/optimizer/BindRelExpr.cpp

```diff
@@ -717,7 +717,7 @@ static ItemExpr *intersectColumns(const RETDesc &leftTable,
     ItemExpr *leftExpr  = leftTable.getValueId(i).getItemExpr();
     ItemExpr *rightExpr = rightTable.getValueId(i).getItemExpr();
     BiRelat *compare = new (bindWA->wHeap())
-      BiRelat(ITM_EQUAL, leftExpr, rightExpr);
+      BiRelat(ITM_EQUAL, leftExpr, rightExpr, TRUE);
```

Comment:
My hope is that hash and merge join are carefully implemented to make this type of equals predicate with the "special nulls" semantics work as an equi-join predicate. If we hash/sort NULL values as regular values, and preserve the "special nulls" semantics for the actual comparison that's done in the hash/merge join, it might work just fine, and hopefully your example proves that this is indeed how it's implemented. The easiest way to check this is probably an EXPLAIN and a SHOWPLAN. If you want to send me this info for the example you added to the regression test (the one that includes a NULL value), I can take a look.